File: app/src/main/java/com/water/datasample/helper/SQLiteHelper.java
package com.water.datasample.helper;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

public class SQLiteHelper extends SQLiteOpenHelper {
    private static String _dname = "ds.db";// 数据库名称
    private static SQLiteDatabase.CursorFactory _factory = null;
    private static int _version = 1;

    private SQLiteDatabase _sd; // 声明SQLiteDatabase对象
    private Context _ctx;

    public SQLiteHelper(Context ctx) {

        this(ctx, 1);
    }

    public SQLiteHelper(Context ctx, int version) {
        this(ctx, _dname, null, version);
    }

    public SQLiteHelper(Context ctx, String database) {
        this(ctx, database, _factory, _version);
    }

    public SQLiteHelper(Context ctx, String database, int version) {
        this(ctx, database, _factory, version);
    }

    public SQLiteHelper(Context ctx, String database,
                        SQLiteDatabase.CursorFactory factory,
                        int version) {
        super(ctx, database, factory, version);
        _ctx = ctx;
        _dname = database;
        _factory = factory;
        _version = version;
    }

    // 数据库第一次创建时被调用
    @Override
    public void onCreate(SQLiteDatabase db) {
        _sd = db;
        //创建表
        String sql = "CREATE TABLE student(" +
                "id INTEGER PRIMARY KEY AUTOINCREMENT," +
                "name text, sex text)";
        _sd.execSQL(sql);
    }
    // 插入
    public void insert(String tablename, ContentValues values){
        if (_sd == null || !_sd.isOpen())
            _sd = getWritableDatabase();

        long count = _sd.insert(tablename,null,values);
        _sd.close();
    }

    // 查询
    public Cursor query(String tablename, String[] columns,
                        String limit){
        if (_sd == null || !_sd.isOpen())
            _sd = getWritableDatabase();

        return _sd.query(tablename, columns,null,
                null,null,null,null, limit);
    }

    // 删除
    public void delete(String tablename, int id){
        if (_sd == null || !_sd.isOpen())
            _sd = getWritableDatabase();

        _sd.delete(tablename,"id=?", new String[]{String.valueOf(id)});
        _sd.close();
    }

    public long getCount(String tablename){
        if (_sd == null || !_sd.isOpen())
            _sd = getWritableDatabase();

        Cursor cursor =  _sd.query(tablename, new String[]{"COUNT(*)"},null,
                null,null,null,null);
        cursor.moveToFirst();
        long ret = cursor.getLong(0);
        cursor.close();
        return ret;
    }

    public void execSQL(String sql)
    {
        if (_sd == null || !_sd.isOpen())
            _sd = getWritableDatabase();

        _sd.execSQL(sql);
    }

    public boolean IsTableExist(String tablename) {
        if (_sd == null || !_sd.isOpen())
            _sd = getWritableDatabase();

        Cursor cursor = _sd.rawQuery("SELECT count(*) FROM sqlite_master " +
                "WHERE type='table' AND name='" + tablename + "'", null);

        boolean result = cursor.getInt(0)==0;

        cursor.close();

        return result;
    }

    // 关闭数据库
    public void close(){
        if (_sd != null || _sd.isOpen())
            _sd.close();
    }

    // 软件版本号发生改变时调用
    @Override
    public void onUpgrade(SQLiteDatabase sd, int i, int i1) {
        _sd = sd;
        _sd.execSQL("ALTER TABLE student ADD age INTEGER");
    }

    public void deleteDb(String tablename) {

        _ctx.deleteDatabase(tablename);
    }
}

File: app/src/main/java/com/github/androidhappyclub/datasample/helper/SQLiteHelper.kt
/*
 * MIT License
 *
 * Copyright (c) 2023 AndroidHappyClub
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.github.androidhappyclub.datasample.helper

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteDatabase.CursorFactory
import android.database.sqlite.SQLiteOpenHelper

class SQLiteHelper private constructor(
    private val context: Context,
    dataName: String? = null,
    factory: CursorFactory? = mFactory,
    version: Int = mVersion
) : SQLiteOpenHelper(context, dataName, factory, version) {
    private var _sd: SQLiteDatabase? = null // 声明SQLiteDatabase对象

    constructor(context: Context) : this(context, mDataName, null, mVersion)
    constructor(context: Context, version: Int) : this(context, mDataName, null, version)
    constructor(context: Context, database: String, version: Int) : this(
        context,
        database,
        mFactory,
        version
    )

    init {
        mDataName = dataName
        mFactory = factory
        mVersion = version
    }

    // 数据库第一次创建时被调用
    override fun onCreate(db: SQLiteDatabase) {
        _sd = db
        //创建表
        val sql = "CREATE TABLE student(id INTEGER PRIMARY KEY AUTOINCREMENT,name text,sex tex)"
        _sd?.execSQL(sql)
    }

    // 插入
    fun insert(tableName: String, values: ContentValues) {
        if (_sd == null || !_sd!!.isOpen) _sd = writableDatabase
        _sd!!.insert(tableName, null, values)
        _sd!!.close()
    }

    // 查询
    fun query(tableName: String, columns: Array<String>, limit: String): Cursor {
        if (_sd == null || !_sd!!.isOpen)
            _sd = writableDatabase
        return _sd!!.query(
            tableName, columns, null,
            null, null, null, null, limit
        )
    }

    // 删除
    fun delete(tableName: String, id: Int) {
        if (_sd == null || !_sd!!.isOpen) _sd = writableDatabase
        _sd!!.delete(tableName, "id=?", arrayOf(id.toString()))
        _sd!!.close()
    }

    fun getCount(tableName: String?): Long {
        if (_sd == null || !_sd!!.isOpen) _sd = writableDatabase
        val cursor = _sd!!.query(
            tableName, arrayOf("COUNT(*)"), null,
            null, null, null, null
        )
        cursor.moveToFirst()
        val ret = cursor.getLong(0)
        cursor.close()
        return ret
    }

    fun execSQL(sql: String?) {
        if (_sd == null || !_sd!!.isOpen) _sd = writableDatabase
        _sd!!.execSQL(sql)
    }

    fun isTableExist(tableName: String): Boolean {
        if (_sd == null || !_sd!!.isOpen) _sd = writableDatabase
        val cursor = _sd!!
            .rawQuery(
                "SELECT count(*) FROM sqlite_master WHERE type= 'table' AND name= 'tableName'",
                null
            )
        val result = cursor.getInt(0) == 0
        cursor.close()
        return result
    }

    // 关闭数据库
    override fun close() {
        _sd?.close()
    }

    // 软件版本号发生改变时调用
    override fun onUpgrade(sd: SQLiteDatabase, i: Int, i1: Int) {
        _sd = sd
        _sd!!.execSQL("ALTER TABLE student ADD age INTEGER")
    }

    fun deleteDb(tableName: String) {
        context.deleteDatabase(tableName)
    }

    companion object {
        private var mDataName: String? = "ds.db" // 数据库名称
        private var mFactory: CursorFactory? = null
        private var mVersion = 1
    }
}